File: electron/i18n.js
'use strict';

/**
 * Lightweight i18n module for Ask Ubuntu Electron renderer.
 *
 * Usage:
 *   await initI18n();
 *   t('welcome.heading')  // → "Ask Ubuntu"
 *   t('status.downloading', { model: 'llama3' })  // → "Downloading llama3…"
 */

let _strings = {};
let _locale = 'en';

/**
 * Initialize i18n by loading locale strings from the main process via IPC.
 */
async function initI18n() {
  try {
    _locale = await window.electronAPI.getLocale();
  } catch (_) {
    _locale = 'en';
  }
  try {
    _strings = await window.electronAPI.getLocaleStrings();
  } catch (_) {
    _strings = {};
  }
}

/**
 * Look up a translated string by key with optional placeholder interpolation.
 *
 * Placeholders use {name} syntax: t('status.downloading', { model: 'llama3' })
 * For plural forms, use pipe-separated singular|plural with {count}:
 *   "tool_calls.summary": "{count} tool call|{count} tool calls"
 *
 * @param {string} key
 * @param {Object} [params]
 * @returns {string}
 */
function t(key, params) {
  let text = _strings[key] || key;

  // Handle simple plural: "singular|plural" split by pipe
  if (text.includes('|') && params && 'count' in params) {
    const parts = text.split('|');
    text = params.count === 1 ? parts[0] : parts[1];
  }

  if (params) {
    for (const [k, v] of Object.entries(params)) {
      text = text.replace(new RegExp(`\\{${k}\\}`, 'g'), String(v));
    }
  }

  return text;
}

/**
 * Get the resolved locale code.
 * @returns {string}
 */
function getResolvedLocale() {
  return _locale;
}

// ── Intl-based formatters ─────────────────────────────────────────────────────

/**
 * Format a number with locale-aware thousand separators.
 * @param {number} n
 * @returns {string}
 */
function formatNumber(n) {
  return new Intl.NumberFormat(_locale).format(n);
}

/**
 * Format temperature — °F for en-US, °C everywhere else.
 * @param {number} celsius
 * @returns {string}
 */
function formatTemperature(celsius) {
  if (_locale === 'en' || _locale === 'en-US') {
    const f = celsius * 9 / 5 + 32;
    return `${Math.round(f)}°F`;
  }
  return `${Math.round(celsius)}°C`;
}

/**
 * Format a time using locale-appropriate 12h/24h convention.
 * @param {Date} date
 * @returns {string}
 */
function formatTime(date) {
  return new Intl.DateTimeFormat(_locale, {
    hour: 'numeric',
    minute: '2-digit',
  }).format(date);
}

/**
 * Format a date using locale-appropriate date order.
 * @param {Date} date
 * @returns {string}
 */
function formatDate(date) {
  return new Intl.DateTimeFormat(_locale, {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(date);
}

/**
 * Format a currency amount.
 * @param {number} amount
 * @param {string} code - ISO 4217 currency code (e.g. 'USD', 'EUR')
 * @returns {string}
 */
function formatCurrency(amount, code) {
  return new Intl.NumberFormat(_locale, {
    style: 'currency',
    currency: code,
  }).format(amount);
}

File: i18n.py
"""
Lightweight i18n module for Ask Ubuntu CLI.

Usage:
    import i18n
    i18n.init()               # detect locale, load strings
    i18n.t('cli.goodbye')     # → "Goodbye!"
    i18n.t('cli.downloading', model='llama3')  # → "Downloading llama3…"
"""

import json
import locale
import os
from pathlib import Path

_strings: dict = {}
_locale_code: str = 'en'


def _detect_locale() -> str:
    """Detect language code from LANG environment variable."""
    lang = os.environ.get('LANG', '')
    # e.g. "es_ES.UTF-8" → "es_ES" → try exact then language-only
    code = lang.split('.')[0]  # strip encoding
    return code if code else 'en'


def _resolve_locale(code: str, locales_dir: Path) -> str:
    """Resolve locale code to an available JSON file.

    Resolution order: exact match (en_GB) → language only (en) → fallback 'en'.
    """
    # Exact match
    if (locales_dir / f'{code}.json').is_file():
        return code
    # Language-only fallback (e.g. es_ES → es)
    lang = code.split('_')[0]
    if lang != code and (locales_dir / f'{lang}.json').is_file():
        return lang
    return 'en'


def _find_locales_dir() -> Path:
    """Locate the locales/ directory, handling snap and dev environments."""
    snap = os.environ.get('SNAP')
    if snap:
        snap_dir = Path(snap) / 'locales'
        if snap_dir.is_dir():
            return snap_dir
    return Path(__file__).resolve().parent / 'locales'


def init(locale_override: str = None):
    """Initialize i18n: detect locale, load base English + locale overlay."""
    global _strings, _locale_code

    locales_dir = _find_locales_dir()

    # Load base English strings
    en_path = locales_dir / 'en.json'
    if en_path.is_file():
        with open(en_path, encoding='utf-8') as f:
            _strings = json.load(f)

    # Determine and resolve locale
    raw_code = locale_override or _detect_locale()
    _locale_code = _resolve_locale(raw_code, locales_dir)

    # Overlay locale-specific strings on top of English base
    if _locale_code != 'en':
        loc_path = locales_dir / f'{_locale_code}.json'
        if loc_path.is_file():
            with open(loc_path, encoding='utf-8') as f:
                _strings.update(json.load(f))

    # Try to set the system locale for number/date formatting
    try:
        locale.setlocale(locale.LC_ALL, '')
    except locale.Error:
        pass


def t(key: str, **kwargs) -> str:
    """Look up a translated string by key, with optional placeholder interpolation.

    Placeholders use {name} syntax: t('cli.downloading', model='llama3')
    For plural forms, use pipe-separated singular|plural with {count}:
        "tool_calls.summary": "{count} tool call|{count} tool calls"
    """
    text = _strings.get(key, key)

    # Handle simple plural: "singular|plural" split by pipe
    if '|' in text and 'count' in kwargs:
        parts = text.split('|', 1)
        text = parts[0] if kwargs['count'] == 1 else parts[1]

    if kwargs:
        try:
            text = text.format(**kwargs)
        except (KeyError, IndexError):
            pass

    return text


def get_locale() -> str:
    """Return the resolved locale code."""
    return _locale_code


def format_number(n) -> str:
    """Format a number with locale-aware thousand separators."""
    try:
        return locale.format_string('%g', n, grouping=True)
    except (ValueError, TypeError):
        return str(n)


def format_temperature(celsius: float) -> str:
    """Format temperature, using Fahrenheit for US locale, Celsius otherwise."""
    if _locale_code in ('en', 'en_US'):
        f = celsius * 9 / 5 + 32
        return f'{f:.0f}\u00b0F'
    return f'{celsius:.0f}\u00b0C'


def format_bytes_localized(n: int) -> str:
    """Format byte count with locale-aware number formatting."""
    if n == 0:
        return '0 B'
    units = ['B', 'KB', 'MB', 'GB', 'TB']
    i = 0
    val = float(n)
    while val >= 1024 and i < len(units) - 1:
        val /= 1024
        i += 1
    formatted = locale.format_string('%.1f', val, grouping=True) if i > 1 else locale.format_string('%.0f', val, grouping=True)
    return f'{formatted} {units[i]}'
